src/main/scala/org/coroutines/AstCanonicalization.scala

```diff
@@ -122,6 +121,48 @@ trait AstCanonicalization[C <: Context] {
       (decls, q"$localvarname")
     case q"$selector[..$tpts](...$paramss)" if tpts.length > 0 || paramss.length > 0 =>
       // application
+
+      // We first need to see which parameters, if any, are by-name. Those that are
+      // should not be canonicalized.
+
+      // Maps to the parameter lists, saying whether or not they are by-name.
+      val byNameParams: immutable.Seq[immutable.Seq[Boolean]] = {
+        // Check that `selector` has a non-trivial symbol. If it doesn't, we assume
+        // that there were no by-name parameters.
+        if (selector.symbol != null && selector.symbol != NoSymbol) {
+          val paramLists = selector.symbol.asMethod.paramLists
+
+          // This value does not contain repeated parameters.
+          val noRepeatedParamsSeq = paramLists.map { currentParamList =>
+            currentParamList.map { param => param.asTerm.isByNameParam }
+          }
+
+          // Check that parameters were both passed in to and requested by the method.
+          if (paramss.length > 0 && noRepeatedParamsSeq.length > 0) {
+            // Initially, we assume that no parameters were repeated.
+            var paramsSeq = noRepeatedParamsSeq
+
+            /** If the number of passed-in parameters is greater than the number of
+             *  parameters needed by the method, then we assume that there are repeated
+             *  parameters.
+             *
+             *  This problem is fixed by appending the last value in
+             *  `noRepeatedParamsSeq` to `paramsSeq`. This is done until the number of
+             *  parameters are equal.
+             */
+            while (paramss(0).length > paramsSeq(0).length) {
+              val newHead = paramsSeq(0) :+ paramsSeq(0).last
+              val newTail = paramsSeq.tail
+              paramsSeq = newHead :: newTail
+            }
+            paramsSeq
+          } else {
+            noRepeatedParamsSeq
+          }
+        } else {
+          immutable.Seq.fill(1, paramss(0).length)(false)
+        }
+      }
```

Comment:
You should not inspect the method symbol and match the parameter trees to the method type information to figure out which is which - you're reproducing complex typechecker logic by doing that, and not covering all the edge cases.

Instead, see this:

http://stackoverflow.com/questions/29757584/handling-by-name-parameters-in-scala-macro

The `ByNameParamClass` is a dummy class that denotes that a parameter is by-name. Unfortunately, this dummy class is only used to wrap the type of the by-name parameter, but it does not appear in any way in an actual application. This is a flaw in the reflection API, as it does not expose enough information.

We should have the following behavior:

1. If the lengths of the argument lists at the callsite is equal to the length of the parameter lists at the method definition, AND, the respective argument types exactly match respective parameter types, then use that information to figure out what is by-name and what is not. This already covers a majority of cases.

2. Otherwise, do not try to patch the argument lists, and immediately report an error - such a method application cannot be used inside a coroutine definition. Workaround for users is to introduce some sort of bridge method.